File: sample.java
// Time Complexity :
// Space Complexity :
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
//Palindrome Linked List
//Time Complexity -> O(n)
//Space Complexity -> O(1)
class Solution {
    public boolean isPalindrome(ListNode head) {
          if(head == null || head.next == null) return true; //a list with no or only one element is a palindrome
          ListNode mid = getMid(head);
          ListNode list1 = head;
          ListNode curr = head;
          while(curr.next!=mid){
              curr = curr.next;
          }
          curr.next = null;
          ListNode list2 = mid;
          list2 = reverseList(list2);
          while(list1!=null && list2!=null){
              if(list1.val != list2.val){
                  return false;
              }  
              list1 = list1.next;
              list2 = list2.next;

          }

        return true;
    }

    private ListNode getMid(ListNode head){
        if(head == null || head.next == null) return head;
        ListNode slow = head;
        ListNode fast = head;
        while(fast!=null && fast.next!=null){
            slow = slow.next;
            fast = fast.next.next;
        }

        return slow;
    }

    private ListNode reverseList(ListNode head){
        if(head == null || head.next == null) return head;
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode prev = null;
        ListNode curr = head;
        ListNode temp = head.next;
        while(temp!=null){
            curr.next = prev;
            prev = curr;
            curr = temp;
            temp = temp.next;
        }
        curr.next = prev;
        return curr;

    }

    
}

//2nd Height Balanced Tree
//Time Complexity -> O(n)
//Space Complexity -> O(h)
class Solution {
    public boolean isBalanced(TreeNode root) {

        if(root == null) return true;

        int leftHeight = height(root.left);
        int rightHeight = height(root.right);
        if(Math.abs(leftHeight-rightHeight)<=1 && isBalanced(root.right) && isBalanced(root.left)){
            return true;
        }
        else{
            return false;
        }

        
        
    }

    private int height(TreeNode root){
        if(root == null) return 0;

        int leftHeight = height(root.left);
        int rightHeight = height(root.right);

        return 1 + Math.max(leftHeight,rightHeight);
        
    }
}

File: Sample.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    
    """
    Description: Given a singly linked list, determine if it is a palindrome?
    
    Time Complexicity: O(n), n -> number of items in given linked-list
    Space Complexicity: O(n) {I think because I am creating another linked-list (mirror)}
    
    Approach:
    1. find the middle node using 2-pointers
    2. create another linked-list by reversing the list found after middle (slow pointer)
    3. match values from the original half and newly created linked-list to check palindrome
    
    """
    
    def isPalindrome(self, head: ListNode) -> bool:
        
        if head == None: return True
        slow = head; fast = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
            
        mirror = slow
        mirror = self.reverse(mirror)
        
        while mirror:
            if mirror.val != head.val:
                return False
            mirror = mirror.next
            head = head.next
            
        return True

    # Function to get reverse linked list
    def reverse(self, head):
        prev = None
        while head:
            temp = head
            head = head.next
            temp.next = prev
            prev = temp
        
        return prev

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    
    """
    Description: Given a binary tree, determine if it is height-balanced.
    
    Time Complexicity: O(n), n -> number of items in the given tree
    Space Complexicity: O(1)
    
    Approach:
    
    1. Define a function height to find height of each subtree (use recursively)
    2. check each left and right roots have consistently difference of heights <= 1 (recursively)
    
    """
    
    def isBalanced(self, root: TreeNode) -> bool:
        
        if root == None: return True
        
        return self.isBalanced(root.left) and self.isBalanced(root.right) and abs(self.height(root.left) - self.height(root.right)) <= 1
        
    def height(self, root):

        """
        Function to get maximum height of the subtree (recursively)
        """

        if root == None: return 0
        else:
            return max(self.height(root.left), self.height(root.right)) + 1
